index.d.ts

```diff
@@ -1,96 +1,640 @@
-    type CustomOrigin = (
-        requestOrigin: string | undefined,
-        callback: (err: Error | null, allow?: boolean) => void
-    ) => void;
-
-    interface CorsOptions {
-        origin?: boolean | string | RegExp | (string | RegExp)[] | CustomOrigin;
-        methods?: string | string[];
-        allowedHeaders?: string | string[];
-        exposedHeaders?: string | string[];
-        credentials?: boolean;
-        maxAge?: number;
-        preflightContinue?: boolean;
-        optionsSuccessStatus?: number;
-    }
-
-    class Route {
-        callOptions: any
-        cors: CorsOptions
-        etag: boolean | "weak" | "strong" | Function
-        hasWhitelist: boolean
-        logging: boolean
-        mappingPolicy: string
-        middlewares: Array<Function>
-        onBeforeCall?: onBeforeCall
-        onAfterCall?: onAfterCall
-        opts: any
-        path: string
-        whitelist: Array<string>
-    }
-
-    type onBeforeCall = (ctx: Context, route: Route, req: IncomingMessage, res: ServerResponse) => void
-    type onAfterCall = (ctx: Context, route: Route, req: IncomingMessage, res: ServerResponse, data: any) => void
-
-    class IncomingRequest extends IncomingMessage {
-        $action: ActionSchema
-        $alias: Alias
-        $ctx: Context
-        $endpoint: ActionEndpoint
-        $next: any
-        $params: any
-        $route: Route
-        $service: Service
-        $startTime: Array<number>
-    }
-
-    class GatewayResponse extends ServerResponse {
-        $ctx: Context
-        $route: Route
-        $service: Service
-    }
-
-    const ApiGatewayService: ServiceSchema & { Errors: ApiGatewayErrors, IncomingRequest: IncomingRequest, GatewayResponse: GatewayResponse };
-    export = ApiGatewayService;
+	import {
+		ActionEndpoint,
+		ActionSchema,
+		CallingOptions,
+		Context,
+		Errors,
+		LogLevels,
+		Service,
+		ServiceSchema,
+	} from 'moleculer'
+	import { IncomingMessage, ServerResponse } from 'http'
+
+	// RateLimit
+	type generateRateLimitKey = (req: IncomingMessage) => string
+
+	interface RateLimit {
+		/**
+		 * How long to keep record of requests in memory (in milliseconds).
+		 * @default 60000 (1 min)
+		 */
+		window?: number,
+
+		/**
+		 * Max number of requests during window.
+		 * @default 30
+		 */
+		limit?: number,
+
+		/**
+		 * Set rate limit headers to response.
+		 * @default false
+		 */
+		headers?: boolean,
+
+		/**
+		 * Function used to generate keys.
+		 * @default req => req.headers["x-forwarded-for"] || req.connection.remoteAddress || req.socket.remoteAddress || req.connection.socket.remoteAddress
+		 */
+		key?: generateRateLimitKey,
+
+		/**
+		 * use rate limit Custom Store
+		 * @default MemoryStore
+		 * @see https://moleculer.services/docs/0.14/moleculer-web.html#Custom-Store-example
+		 */
+		StoreFactory?: MemoryStore
+	}
+
+	class MemoryStore {
+		constructor(window: number, opts: RateLimit)
```

Comment:
The constructor in the docs shows this as `window` (as indicated here) but the actual code refers to this as `clearPeriod`.  Additionally, the docs don't end up using `window` in their example and are also referencing `clearPeriod`.

That might not be indicative of a problem with this PR and possibly `window` is the better name for this parameter, but there is a lack of consistency between the actual code, the docs, and this PR right now.

@icebob Is the preferred parameter name here `window` or `clearPeriod`?

Reply:
Let's use the `clearPeriod` because it is in the source code. I'm fixing it in the docs.